### ISSUE TITLE
Add distance of remaining path to Navigation Agent

### ISSUE THREAD

**Issue by elvisish** [godotengine/godot-proposals#8296]
### Describe the project you are working on

Walking simulator.

### Describe the problem or limitation you are having in your project

Currently, navigation agents have a way of returning the direct distance to the target and the current navigation path array, but no way of getting the length of the remaining distance *along* the navigation path. It might be possible to get something close, but it's cumbersome:

```swift
sum_navpath(nav.get_current_navigation_path())

func sum_navpath(arr:Array):
	var current_index := 0
	var result = 0
	for i in arr:
		if current_index < nav.get_current_navigation_path_index():
			current_index += 1
			continue
		result+=i.length()
	if arr.size() > 0:
		result-= arr[arr.size() - 1].length()
	return result
```

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

A function that returns the remaining path length.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

A function in Navigation Agent.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

A navigation agent feature.

### Is there a reason why this should be core and not an add-on in the asset library?

A navigation agent feature.

**Comment by AThousandShips** [comment]
I don't believe there's any internal tacking of this, so it'd just be a helper function doing the workaround you mention

**Comment by elvisish** [comment]
> I don't believe there's any internal tacking of this, so it'd just be a helper function doing the workaround you mention

I guess I'm proposing that it should be tracked internally, the helper function I wrote doesn't really work properly and it snaps to each path section rather than gradienting along the path towards the target (if a better helper can be written I'd be very interested as that was really the best I could manage)

**Comment by elvisish** [comment]
> And I think the issues you face indicates how it's not necessarily trivial to estimate, there's no way to reliably judge it even without avoidance, as it depends entirely on how the user applies the navigation, since the engine leaves the movement to the user

I think it just needs to be the same distance in meters or whatever Godot uses for transform positions, most users apply navigation by setting velocity, I don't see how this would be an especially different implementation to the existing crow-flies distance to the target (just using the path length rather than the global_transform distances.

**Comment by AThousandShips** [comment]
Well that's not what I mean, I don't mean units, I mean that there's no reliable way to make an estimate of the remaining distance because it depends entirely on factors outside of the navigation code

**Comment by elvisish** [comment]
> Well that's not what I mean, I don't mean units, I mean that there's no reliable way to make an estimate of the remaining distance because it depends entirely on factors outside of the navigation code

Okay, but Unity has this, so I guess however that is done would be just as fine for Godot:  https://docs.unity3d.com/540/Documentation/ScriptReference/NavMeshAgent-remainingDistance.html

> So I fail to see the reason for adding this in core

Because I don't think it can be done easily, reliably or quickly in gdscript (unless there's a way I'm missing, in which case I'd be more than happy to learn).

**Comment by smix8** [comment]
That is adding up the vector lengths between the path points like you already do.

Something that could theoretically be added as a path meta option to toggle on, because most projects will probably never need that so they shouldn't be affected by the performance hit to calculate all len updates.

**Comment by elvisish** [comment]
> That is adding up the vector lengths between the path points like you already do.
> 
> Something that could theoretically be added as a path meta option to toggle on, because most projects will probably never need that so they shouldn't be affected by the performance hit to calculate all len updates.

Yeah, even I don't need it much (just for a one-off situation) so it would definitely make sense to be toggleable. But my way of calculating them doesn't work very well :/

**Comment by tivvit** [comment]
Hi @elvisish, I needed to do the same thing you are doing. Thanks for your code!

It may work well for your case, but you are summing up the lengths in the global coordinates. Here is my code, which sums up the distances of each consecutive coordinates. It may help someone.

```
func sum_navpath(arr: Array):
	var current_index := 0
	var result = 0
	var current_pos = arr[nav.get_current_navigation_path_index()]
	for i in arr:
		if current_index <= nav.get_current_navigation_path_index():
			current_index += 1
			continue
		result += current_pos.distance_to(i)
		current_pos = i
	return result
```

**Comment by elvisish** [comment]
> Hi @elvisish, I needed to do the same thing you are doing. Thanks for your code!
> 
> 
> 
> It may work well for your case, but you are summing up the lengths in the global coordinates. Here is my code, which sums up the distances of each consecutive coordinates. It may help someone.
> 
> 
> 
> ```gdscript
> 
> func sum_navpath(arr: Array):
> 
> 	var current_index := 0
> 
> 	var result = 0
> 
> 	var current_pos = arr[nav.get_current_navigation_path_index()]
> 
> 	for i in arr:
> 
> 		if current_index <= nav.get_current_navigation_path_index():
> 
> 			current_index += 1
> 
> 			continue
> 
> 		result += current_pos.distance_to(i)
> 
> 		current_pos = i
> 
> 	return result
> 
> ```

Oh nice, I couldn't figure out how to make mine work properly, I'll try this and see how it goes, thanks again!

**Comment by Exzircon** [comment]
This was almost exactly what I needed too, but it doesn't seem to calculate the distance from your nav agent to the first nav path index. I just added the distance from the nav agent to the current_pos at the start of the function (had to move the result declaration a line down too)

```python
func sum_nav_path(arr: Array):
	var current_index := 0
	var current_pos = arr[nav.get_current_navigation_path_index()]
	var result = global_position.distance_to(current_pos)
	for i in arr:
		if current_index <= nav.get_current_navigation_path_index():
			current_index += 1
			continue
		result += current_pos.distance_to(i)
		current_pos = i
	return result
```
